### PR TITLE
13914 use read/write locking on Windows

### LINKED ISSUE #13914: Read/Write locking on Windows / FIPS
My company is currently using version 1.0.2 on Windows. We recently discovered a performance issue when running our product in FIPS mode. Debugging proved lock contention within hashing functions, essentially single-threading password validation throughout all threads in our product. 

After some analysis, we concluded that the locking callbacks we provided to OpenSSL through `CRYPTO_set_locking_callback` were insufficient, checking only the `CRYPTO_LOCK` and `CRYPTO_UNLOCK` bits and using a Windows mutex to synchronize.

So we quickly replaced the callbacks to handle the additional `CRYPTO_READ` and `CRYPTO_WRITE` bits and use an [SRWLock](https://docs.microsoft.com/en-us/windows/win32/sync/slim-reader-writer--srw--locks) primitive for synchronization. This resolved the issue and we're shipping a patch ASAP.

Follow-up searching turned up another OpenSSL 1.0.2 user [who had encountered a similar problem](https://mta.openssl.org/pipermail/openssl-users/2015-April/001288.html). Their solution was to ignore the FIPS-related mutexes after enabling FIPS had completed.

This raises some issues with our upgrade path. When FIPS is available/possible in version 3.0 we would like to upgrade, but it seems as though we would need read/write locking support on Windows to avoid this performance issue.

OpenSSL now (since 1.1.1?) implements its own thread synchronization. The [current Windows implementation](https://github.com/openssl/openssl/blob/master/crypto/threads_win.c) uses a `CriticalSection` with no checks for the `CRYPTO_READ` and `CRYPTO_WRITE` bits. It appears that the [threading callbacks have been removed](https://github.com/openssl/openssl/blob/3aa7212e0a4fd1533c8a28b8587dd8b022f3a66f/include/openssl/crypto.h.in#L255), so we would not be able to "specifically ignore" the FIPS mutexes.

Should we anticipate running into the same issue with FIPS on version 3.0?
If so, is there interest (aside from myself) in implementing read/write locking for Windows?

Thanks!

### PR DESCRIPTION
**Pull request by VeeFu**
Use read/write locking on Windows

    Fixes #13914

    The "SRWLock" synchronization primitive is available in Windows Vista
    and later.  CRYPTO_THREAD functions now use SRWLock functions when the
    target operating system supports them.

##### Checklist

- [x] documentation is added or updated
- [x] `test_encoder_decoder_legacy.t` currently hangs. Either #14319 or #14171 will fix.

### CODE REVIEW

File: crypto/threads_win.c
Comment (slontis) [inline crypto/threads_win.c:38]: Just a few NITS here.. (Note the blankline before the if)

````
# ifdef USE_RWLOCK
    CRYPTO_win_rwlock *rwlock;

     if ((lock
````

File: crypto/threads_win.c
Comment (slontis) [inline crypto/threads_win.c:69]: blankline before Acquire

File: crypto/threads_win.c
Comment (slontis) [inline crypto/threads_win.c:94]: not sure the switch and default cases are needed here..
The init sets it up in SRWLOCK_MODE_SHARED
 when it does the zalloc. Something is terribly wrong if it is not one of these 2 states.

**Comment by slontis** [comment]
Should this have a CHANGES entry?

File: crypto/threads_win.c
Comment (paulidale) [inline crypto/threads_win.c:30]: I'd not bother with an enum here, it's a Boolean value.  Simply `int shared` or `int exclusive` is enough.  Keep it simple.

File: crypto/threads_win.c
Comment (paulidale) [reply crypto/threads_win.c:94]: Agreed, just an if statement is enough here.  More so if the enum is removed.

**Review by paulidale — CHANGES_REQUESTED** [review]

File: crypto/threads_win.c
Comment (slontis) [inline crypto/threads_win.c]: Just use 1 and 0 for TRUE and FALSE.

I can only see one instance of TRUE in the codebase and it is related to win32 code that uses BOOL ret;

**Comment by slontis** [comment]
Once #14319 is merged (in a day or so), you should be able to rebase this and remove the Merge branch..

Rewording the commit message of the first commit to just say...

````
Use read/write locking on Windows

Fixes #13914

The "SRWLock" ....
````

File: crypto/threads_win.c
Comment (mattcaswell) [inline crypto/threads_win.c:66]: I would rather make this "nonshared" (or similar name) and only write to it when acquiring a write lock, and then reset it to 0 just prior to unlock. That way we never have to write to this value when acquiring a read lock (with multiple threads all competing to write the same value).

**Comment by mattcaswell** [comment]
> Should this have a CHANGES entry?

Yes - I think a CHANGES.md entry is warranted for this.

**Comment by VeeFu** [comment]
> Once #14319 is merged (in a day or so), you should be able to rebase this and remove the Merge branch..
> 

Yes... the merge commit was a failed attempt to replace the first commit, which was misattributed to my work email address. I'm not quite sure how to fix this. Re-pushing the remote branch comes to mind, but I thought that might lose review history. I do intend to clean this up before merge; I'll need to learn a bit more git before I can, though.

File: crypto/threads_win.c
Comment (VeeFu) [reply crypto/threads_win.c]: Thanks, very good point.

File: crypto/threads_win.c
Comment (mattcaswell) [inline crypto/threads_win.c:90]: Coding style: space after `)` and before `{`

File: crypto/threads_win.c
Comment (mattcaswell) [inline crypto/threads_win.c:95]: Coding style: if one branch of the "if" has `{}` then both branches should.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Comment by t8m** [comment]
> > Once #14319 is merged (in a day or so), you should be able to rebase this and remove the Merge branch..
> 
> Yes... the merge commit was a failed attempt to replace the first commit, which was misattributed to my work email address. I'm not quite sure how to fix this. Re-pushing the remote branch comes to mind, but I thought that might lose review history. I do intend to clean this up before merge; I'll need to learn a bit more git before I can, though.

Do not worry. Just force push with clean commits.

**Comment by mattcaswell** [comment]
#14319 just got merged...

**Comment by VeeFu** [comment]
> #14319 just got merged...

Thanks, I'll rebase, have another go at deleting the first commit (causing the CLA violation) and force-push the branch.

File: crypto/threads_win.c
Comment (mattcaswell) [inline crypto/threads_win.c:94]: Put the else on the same line as }

````
} else {
````

**Review by mattcaswell — APPROVED** [review]
Approved subject to:
- the nit mentioned below being fixed, and
- the CIs passing once this is rebased and the merge commit removed.

**Comment by mattcaswell** [comment]
I guess this PR can be taken out of "draft" now.

File: crypto/threads_win.c
Comment (VeeFu) [reply crypto/threads_win.c]: nit picked.

**Review by mattcaswell — APPROVED** [review]
Reconfirming my approval. The CI failures all appear to be known issues unrelated to this PR.

Ping for second review.

**Comment by slontis** [comment]
Needs to be rebased..

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
It needs a rebase but merging the changes entry shouldn't impact approval.

**Review by paulidale — APPROVED** [review]

**Comment by mattcaswell** [comment]
Pushed! Thank you for your contribution!!

**Pull request closed without merging**
